src/main/java/org/radarbase/appserver/service/transmitter/FcmTransmitter.java

```diff
@@ -0,0 +1,195 @@
+package org.radarbase.appserver.service.transmitter;
+
+import com.google.firebase.ErrorCode;
+import com.google.firebase.messaging.FirebaseMessagingException;
+import com.google.firebase.messaging.MessagingErrorCode;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import org.radarbase.appserver.dto.fcm.FcmUserDto;
+import org.radarbase.appserver.entity.DataMessage;
+import org.radarbase.appserver.entity.Message;
+import org.radarbase.appserver.entity.Notification;
+import org.radarbase.appserver.exception.FcmMessageTransmitException;
+import org.radarbase.appserver.exception.MessageTransmitException;
+import org.radarbase.appserver.service.FcmDataMessageService;
+import org.radarbase.appserver.service.FcmNotificationService;
+import org.radarbase.appserver.service.UserService;
+import org.radarbase.fcm.downstream.FcmSender;
+import org.radarbase.fcm.model.FcmDataMessage;
+import org.radarbase.fcm.model.FcmNotificationMessage;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Component;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+@Slf4j
+@Component
+public class FcmTransmitter implements NotificationTransmitter, DataMessageTransmitter {
+
+    protected static final boolean IS_DELIVERY_RECEIPT_REQUESTED = true;
+    static final int DEFAULT_TIME_TO_LIVE = 2_419_200; // 4 weeks
+
+    protected final transient FcmSender fcmSender;
+    private final transient FcmNotificationService notificationService;
+    private final transient FcmDataMessageService dataMessageService;
+    private final transient UserService userService;
+
+    public FcmTransmitter(
+        @Autowired @Qualifier("fcmSenderProps") FcmSender fcmSender,
+        @Autowired FcmNotificationService notificationService,
+        @Autowired FcmDataMessageService dataMessageService,
+        @Autowired UserService userService
+        ) {
+        this.fcmSender = fcmSender;
+        this.notificationService = notificationService;
+        this.dataMessageService = dataMessageService;
+        this.userService = userService;
+    }
+
+    @Override
+    @SneakyThrows
+    public void send(Notification notification) {
+        try {
+            fcmSender.send(createMessageFromNotification(notification));
+        }  catch (FirebaseMessagingException exc) {
+            handleFcmException(exc, notification);
+        } catch (Exception exc) {
+            throw new FcmMessageTransmitException("Could not transmit a notification through Fcm", exc);
+        }
+    }
+
+    @Override
+    @SneakyThrows
+    public void send(DataMessage dataMessage) {
+        try {
+            fcmSender.send(createMessageFromDataMessage(dataMessage));
+        }  catch (FirebaseMessagingException exc) {
+            handleFcmException(exc, dataMessage);
```

Comment:
please raise FcmMessageTransmitException (encapsulate the FirebaseMessagingException exc) here too

Reply:
I do not think I fully understand. You suggest that the FirebaseMessageingException will also resulti in a FcmMessageTransmitException? That is different from the orgiginal implementation AFAIKS:

In MessageJob:
````

    } catch (FirebaseMessagingException exc) {
      log.error("Error occurred when sending downstream message.", exc);
      // TODO: update the data message status using event
      if (message != null) {
        handleErrorCode(exc.getErrorCode(), message);
        handleFCMErrorCode(exc.getMessagingErrorCode(), message);
      }
    } catch (Exception e) {
      throw new JobExecutionException(e);
    }
```

Reply:
Okay, agreed. Let's keep the original functionality for now. I was just thinking that if the notification was not sent successfully, that should maybe count towards the unsuccessful job so the notification event lifecycle can be updated with failure, but I think we can tackle it in another issue.